File: trunk/OrcScala/src/orc/lib/state/JavaArray.java
//
// JavaArray.java -- Java class JavaArray
// Project OrcScala
//
// $Id$
//
// Copyright (c) 2009 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state;

import java.lang.reflect.Array;
import java.util.HashMap;

import orc.error.runtime.ArityMismatchException;
import orc.error.runtime.BadArrayElementTypeException;
import orc.error.runtime.TokenException;
import orc.lib.state.types.ArrayType;
import orc.types.Type;
import orc.values.sites.TypedSite;
import orc.values.sites.compatibility.Args;
import orc.values.sites.compatibility.EvalSite;

public class JavaArray extends EvalSite implements TypedSite {
	private static HashMap<String, Class<?>> types = new HashMap<String, Class<?>>();
	static {
		types.put("double", Double.TYPE);
		types.put("float", Float.TYPE);
		types.put("long", Long.TYPE);
		types.put("int", Integer.TYPE);
		types.put("short", Short.TYPE);
		types.put("byte", Byte.TYPE);
		types.put("char", Character.TYPE);
		types.put("boolean", Boolean.TYPE);
	}

	@Override
	public Object evaluate(final Args args) throws TokenException {
		if (args.size() == 1) {
			return Array.newInstance(Object.class, args.intArg(0));
		} else if (args.size() == 2) {
			final Class<?> type = types.get(args.stringArg(1));
			if (type == null) {
				throw new BadArrayElementTypeException(args.stringArg(0));
			}
			return Array.newInstance(type, args.intArg(0));
		} else {
			throw new ArityMismatchException(2, args.size());
		}
	}

	@Override
	public Type orcType() {
		return ArrayType.getBuilder();
	}
}

File: OrcScala/src/orc/lib/state/JavaArray.scala
//
// JavaArray.java -- Java class JavaArray
// Project OrcScala
//
// Copyright (c) 2016 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state

import java.lang.reflect.{Array => JArray}

import orc.{DirectInvoker, OrcRuntime}
import orc.error.runtime.{ArityMismatchException, BadArrayElementTypeException}
import orc.lib.state.types.ArrayType
import orc.types.Type
import orc.util.ArrayExtensions.{Array1, Array2}
import orc.values.sites.{
  TargetClassAndArgumentClassSpecializedInvoker,
  TotalSiteBase,
  TypedSite
}

object JavaArray extends TotalSiteBase with TypedSite {
  override val inlinable = true
  private val types: Map[String, Class[_]] = Map(
    ("double", java.lang.Double.TYPE),
    ("float", java.lang.Float.TYPE),
    ("long", java.lang.Long.TYPE),
    ("int", java.lang.Integer.TYPE),
    ("short", java.lang.Short.TYPE),
    ("byte", java.lang.Byte.TYPE),
    ("char", java.lang.Character.TYPE),
    ("boolean", java.lang.Boolean.TYPE),
  )

  def getInvoker(runtime: OrcRuntime, args: Array[AnyRef]) = {
    args.size match {
      case 1 =>
        invoker(this, args: _*) { (_, args) =>
          val Array1(i: Number) = args
          JArray.newInstance(classOf[AnyRef], i.intValue())
        }
      case 2 =>
        invoker(this, args: _*) { (_, args) =>
          val Array2(i: Number, s: String) = args
          val tpe: Class[_] = types.getOrElse(s, {
            throw new BadArrayElementTypeException(s)
          })
          JArray.newInstance(tpe, i.intValue())
        }
      case _ =>
        new TargetClassAndArgumentClassSpecializedInvoker(this, args)
        with DirectInvoker {
          @throws[Throwable]
          def invokeDirect(target: AnyRef, arguments: Array[AnyRef]): AnyRef = {
            throw new ArityMismatchException(2, args.size)
          }
        }
    }
  }

  override def orcType(): Type = ArrayType.getBuilder
}
